nifi-toolkit/nifi-toolkit-tls/src/main/java/org/apache/nifi/toolkit/tls/service/TlsCertificateSigningRequestPerformer.java

```diff
@@ -0,0 +1,144 @@
+    }
+
+    public TlsCertificateSigningRequestPerformer(Supplier<HttpClientBuilder> httpClientBuilderSupplier, String caHostname, String dn, String token, int port, TlsHelper tlsHelper) {
+        this.httpClientBuilderSupplier = httpClientBuilderSupplier;
+        this.caHostname = caHostname;
+        this.dn = dn;
+        this.token = token;
+        this.port = port;
+        this.tlsHelper = tlsHelper;
+    }
+
+    public static String getDn(String hostname) {
+        return "CN=" + hostname + ",OU=NIFI";
+    }
+
+    /**
+     * Submits a CSR to the Certificate authority, checks the resulting hmac, and returns the chain if everything succeeds
+     *
+     * @param objectMapper for serialization
+     * @param keyPair      the keypair to generate the csr for
+     * @throws IOException if there is a problem during the process
+     * @returnd the resulting certificate chain
+     */
+    public X509Certificate[] perform(ObjectMapper objectMapper, KeyPair keyPair) throws IOException {
+        try {
+            List<X509Certificate> certificates = new ArrayList<>();
+
+            HttpClientBuilder httpClientBuilder = httpClientBuilderSupplier.get();
+            SSLContextBuilder sslContextBuilder = SSLContextBuilder.create();
+
+            // We will be validating that we are talking to the correct host once we get the response's hmac of the token and public key of the ca
+            sslContextBuilder.loadTrustMaterial(null, new TrustSelfSignedStrategy());
+            httpClientBuilder.setSSLSocketFactory(new TlsCertificateAuthorityClientSocketFactory(sslContextBuilder.build(), caHostname, certificates));
+
+            String jsonResponseString;
+            int responseCode;
+            try (CloseableHttpClient client = httpClientBuilder.build()) {
+                JcaPKCS10CertificationRequest request = tlsHelper.generateCertificationRequest(dn, keyPair);
+                TlsCertificateAuthorityRequest tlsCertificateAuthorityRequest = new TlsCertificateAuthorityRequest(tlsHelper.calculateHMac(token, request.getPublicKey()),
+                        tlsHelper.pemEncodeJcaObject(request));
+
+                HttpPost httpPost = new HttpPost();
+                httpPost.setEntity(new ByteArrayEntity(objectMapper.writeValueAsBytes(tlsCertificateAuthorityRequest)));
+
+                try (CloseableHttpResponse response = client.execute(new HttpHost(caHostname, port, "https"), httpPost)) {
+                    jsonResponseString = IOUtils.toString(new BoundedInputStream(response.getEntity().getContent(), 1024 * 1024), StandardCharsets.UTF_8);
+                    responseCode = response.getStatusLine().getStatusCode();
+                }
+            }
+
+            if (responseCode != Response.SC_OK) {
+                throw new IOException(RECEIVED_RESPONSE_CODE + responseCode + " with payload " + jsonResponseString);
```

Comment:
Updated PR to catch exception and wrap in very generic message on server side.  As far as response goes, if you look at TlsCertificateAuthorityServiceHandler, it sends back an enumerated set of error messages that shouldn't reveal too much to a bad guy.